portfolio/src/main/java/com/google/sps/data/GenreRecommender.java

```diff
@@ -69,4 +79,73 @@ public GenreRecommender(List<Book> books) {
     }
     return matches;
   }
+
+  /**
+   * Takes in a set of genres and returns an ImmutableMap
+   * with books and their relevance score
+   *
+   * @param genres: Set of genres (as Strings)
+   * @return ImmutableMap with {Book:relevanceScore}
+   */
+  private ImmutableMap<Book, Integer> getBooksWithScores(Set<String> genres) {
+    ImmutableMap.Builder<Book, Integer> bookToScore = new ImmutableMap.Builder<Book, Integer>();
+    Set<Book> booksToConsider = new HashSet<Book>();
+    for (String genre : genres) {
+      booksToConsider.addAll(genreToBooks.get(genre));
+    }
+
+    for (Book book : booksToConsider) {
+      // assign score based on relevance:
+      int score = 0;
+      for (String genre : getGenres(book)) {
+        if (genres.contains(genre)) {
+          score += MATCH_SCORE;
+        }
+      }
+      bookToScore.put(book, score);
+    }
+    return bookToScore.build();
+  }
+
+  /**
+   * Returns the top N matches of a book by Genres
+   *
+   * @param book: Book object for recommendations
+   * @param n: Number of books to be returned
+   * @return List of n books sorted most-> least recommended
+   */
+  public List<Book> getTopNMatches(Book originalBook, int n) {
+    relevanceScoreMap = getBooksWithScores(originalBook.genre());
+    List<Book> topNBooks = new ArrayList<Book>();
+
+    if (n < 1) {
+      return topNBooks;
+    }
+    for (Book book : relevanceScoreMap.keySet()) {
+      if (book.equals(originalBook)) {
+        continue;
+      }
+      if (topNBooks.size() < n) {
+        addInPosition(topNBooks, book); // add it in the right place
+      } else if (relevanceScoreMap.get(book) > relevanceScoreMap.get(topNBooks.get(n - 1))) {
+        addInPosition(topNBooks.subList(0, n - 1), book);
+      }
+    }
+    return topNBooks;
+  }
+
+  private void addInPosition(List<Book> list, Book book) {
```

Comment:
I think we could use Collections.sort instead! https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#sort(java.util.List)

According to the documentation, "If the input array is nearly sorted, the implementation requires approximately n comparisons."

So it sounds like this would run in O(N) in this scenario, meaning it's no worse than addInPosition() and therefore less code for us!

e.g. Collections.sort(topNBooks, Collections.reverseOrder());